### PR TITLE
Review branch

### PR DESCRIPTION
**Pull request by kalehrishi**

### CODE REVIEW

**Comment by kalehrishi** [comment]
branch name should be related to activity you are performing inside that branch. Usually its a feature that you are adding. In this case it could be `email-validator`

**Comment by kalehrishi** [comment]
Commit messages should be meaningful. `Ok` and `Done` doesn't make any sense. It should explain what change you have made within that commit. For example `Added test cases for Email.validate function` or `Implemented Email.validate function.

File: app/Email.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace emailApp;
```
Comment (kalehrishi) [inline app/Email.php:3]: Namespace name should start with caps. `EmailApp`

File: app/Email.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+class Email {
+	function validateEmail($email) {
```
Comment (kalehrishi) [inline app/Email.php:6]: function name should be plain verb. You should use verb+noun if you are performing activity on something other than its class.  Repeating class name in function name is not good. You can simply name function `validate`

File: app/Email.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+		$regex = '/^[_a-z0-9-]+(\.[_a-z0-9-]+)*@[a-z0-9-]+(\.[a-z0-9-]+)*(\.[a-z]{2,4})$/';
+
+		if (!preg_match($regex, $email)) {
+			throw new \InvalidArgumentException("PLEASE ENTER VALID EMAIL ADDRESS.....");
```
Comment (kalehrishi) [inline app/Email.php:11]: Never use all caps in error message. All caps is considered as yelling. http://www.newrepublic.com/article/117390/netiquette-capitalization-how-caps-became-code-yelling We should write messages in formal English.  `....` isn't a formal punctuation.

File: app/Email.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+		} else {
+			return 1;
```
Comment (kalehrishi) [inline app/Email.php:13]: It should return `boolean`. Comparing `Int` vs comparing `Boolean`. Boolean comparison is always cheap.

File: app/Email.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+		}
+	}
+}
```
Comment (kalehrishi) [inline app/Email.php:16]: Ending a file with newline is good.

File: app/Email.php (hunk 2 above)
Comment (kalehrishi) [inline app/Email.php:6]: There should be comment explaining purpose of function is phpdocs format. http://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_phpDocumentor.howto.pkg.html

File: composer.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+"autoload": {
+        "psr-4": {"emailApp\\": "app/"}
```
Comment (kalehrishi) [inline composer.json:3]: Correct namespace name. as explained in previous comment.

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+"autoload": {
```
Comment (kalehrishi) [inline composer.json:2]: Rest of parameters explaining application are missing. Refer to composer format. https://getcomposer.org/doc/04-schema.md

File: tests/EmailTest.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<?php
+namespace emailApp;
+include dirname(__FILE__)."/../vendor/autoload.php";
```
Comment (kalehrishi) [inline tests/EmailTest.php:3]: You should pass autoload using bootstrap argument to phpunit.

File: tests/EmailTest.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	
+	/**
+	 *@test
+	 */
+	public function it_can_validate_email() {
+		$object = new Email();
+		$result = $object->validateEmail("shantanu3687@gmail.com");
+		$this->assertEquals($result, 1);
+	}
+    /**
```
Comment (kalehrishi) [inline tests/EmailTest.php:15]: Incorrect indentation.

File: tests/EmailTest.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+namespace emailApp;
+include dirname(__FILE__)."/../vendor/autoload.php";
+
+class EmailTest extends \PHPUnit_Framework_TestCase {
+	
+	/**
+	 *@test
+	 */
+	public function it_can_validate_email() {
+		$object = new Email();
```
Comment (kalehrishi) [inline tests/EmailTest.php:11]: Object name should be valid name `$object` is not a good name. Here you can use `$email`.

File: tests/EmailTest.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
+<?php
+namespace emailApp;
+include dirname(__FILE__)."/../vendor/autoload.php";
+
+class EmailTest extends \PHPUnit_Framework_TestCase {
+	
+	/**
+	 *@test
+	 */
+	public function it_can_validate_email() {
```
Comment (kalehrishi) [inline tests/EmailTest.php:10]: Test name doesn't make any sense. Which type of email are you testing?

File: tests/EmailTest.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	public function it_can_validate_email_starting_with_number() {
+		$object = new Email();
+		$result = $object->validateEmail("111shantanu3687@gmail.com");
+		$this->assertEquals($result, 1);
+	}
+	/**
+	 *@test
+	 * @expectedException InvalidArgumentException
+	 */
+	public function it_should_throgh_InvalidArgumentException_for_email_having_more_number_of_at_the_rate_sign() {
```
Comment (kalehrishi) [inline tests/EmailTest.php:28]: Be careful about spelling. Its `throw` not `throgh`

Comment (kalehrishi) [inline tests/EmailTest.php:28]: `@` is called `at sign` https://en.wikipedia.org/wiki/At_sign

In test case title you are saying more than one at signs. but in actual test case you are passing just one. and more dots.

File: tests/EmailTest.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	 *
+	 */
+	public function it_can_validate_email_starting_with_number() {
+		$object = new Email();
+		$result = $object->validateEmail("111shantanu3687@gmail.com");
+		$this->assertEquals($result, 1);
+	}
+	/**
+	 *@test
+	 * @expectedException InvalidArgumentException
```
Comment (kalehrishi) [inline tests/EmailTest.php:26]: You should also test for `expectedMessage`

File: tests/EmailTest.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+	 */
+	public function it_should_throgh_InvalidArgumentException_for_email_starting_with_special_character() {
+		$object = new Email();
+		$object->validateEmail("$###shantanu3687@gmail.com");
+	}
+    /**
+	 *@test
+	 * @expectedException InvalidArgumentException
+	 */
+	public function it_should_throgh_InvalidArgumentException_for_email_which_does_not_having_at_the_rate_sign() {
```
Comment (kalehrishi) [inline tests/EmailTest.php:44]: Correct grammar is `does not contain`

File: tests/EmailTest.php [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+		$object = new Email();
+		$object->validateEmail("shantanu3687gmail.com");
+	}
+    /**
+	 *@test
+	 * @expectedException InvalidArgumentException
```
Comment (kalehrishi) [inline tests/EmailTest.php:50]: Use two spaces for indentation. Do not use tabs.

**Comment by kalehrishi** [comment]
`composer.lock` file is not committed.

File: tests/EmailTest.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+	 */
+	public function it_should_throgh_InvalidArgumentException_for_email_which_does_not_having_domain_name() {
+		$object = new Email();
+		$object->validateEmail("shantanu3687@com");
+	}
+}
```
Comment (kalehrishi) [inline tests/EmailTest.php:56]: Where is test case for empty or non string email id?

File: tests/EmailTest.php (hunk 8 above)
Comment (shantanu7p) [inline tests/EmailTest.php:3]: we did it in both ways .
Using "bootstrap" as well as "include"

**Pull request merged**